board/*/kconfig: Fix boards on kconfig blocklist

MrKevinWeiss opened the pull request:
### Contribution description

This should (eventually) test and pass all non-deprecated boards when using kconfig modeling.

### Testing procedure

Murdock green

### Issues/PRs references

Related to #16875

MrKevinWeiss commented:
Hmmm currently the esps are failing due to RIOT_CI_BUILD changing the dependency resolution.

I thought we can get around it by just importing a `riot-ci-build.config` but, as importing configs can't have any ifs or something, it seems like this will need to be solved by adding a `CONFIG_RIOT_CI_BUILD` that can alter the kconfig dependency resolution.  Maybe this could also be solved with some macro magic or something.

gschorcht commented:
> Hmmm currently the esps are failing due to RIOT_CI_BUILD changing the dependency resolution.

:worried: I am really sorry for this inconvenience. These conditional dependencies were just a trick to cover all optional `esp_*` modules in CI compilation without defining additional virtual boards which would increase again the compile time in Murdock. This became necessary because we noticed several times that after a series of changes provided by different PRs, the compilation suddenly fails when optional modules are enabled because they were not covered when the PRs were compiled in CI. 

Another idea (first approach in PR #12971) was to clone a number of test applications which use these optional modules, but it seemed to be not reasonable to have duplicated code just for compilation tests.  

To be honest, I felt that the virtual boards approach would have been the coolest one and the clearest.

MrKevinWeiss commented:
> I am really sorry for this inconvenience.

Don't be.  This just points out an issue that we should be able to solve.  This is not the only env var that would change the dependency resolution I would image.  Now that I think of it, it should be solved like DEVELHELP.  Let me check it out!

MrKevinWeiss commented:
We will see about this.  I am not so convinced of the `HAVE_RIOT_CI_BUILD_ESP_*` symbols, especially the implementation of `HAVE_RIOT_CI_BUILD_ESP_I2C_HW`.  It would be good to get some feedback from @leandrolanzieri as he figured out that design pattern.  But I can wait for that review. (worst case we flag it and fix it later...)

MrKevinWeiss commented:
Ahh it passed... That seemed almost too easy.

MrKevinWeiss commented:
I will run a full build if the CI is free I guess.

MrKevinWeiss commented:
> I will run a full build if the CI is free I guess.

I spoke too soon :laughing:

MrKevinWeiss commented:
Rebased and cleaned up.

I think this is ready for proper review.  Maybe @leandrolanzieri or @fjmolinas when they are available.

@aabadie I did a small fix to the stm clock in fc538065715b346803e94bd0109898b2b3227e2c but the remaining stm boards will need larger modelling.

fjmolinas commented on boards/esp8266-olimex-mod/Kconfig:
```diff
@@ -13,4 +13,10 @@ config BOARD_ESP8266_OLIMEX_MOD
     select BOARD_COMMON_ESP8266
     select CPU_MODEL_ESP8266_ESP_12X
 
+    select HAVE_RIOT_CI_BUILD_ESP_SW_TIMER
+
+    select MODULE_ESP_LOG_STARTUP if TEST_KCONFIG && RIOT_CI_BUILD
+    select MODULE_ESP_LOG_TAGGED if TEST_KCONFIG && RIOT_CI_BUILD
+    select MODULE_ESP_QEMU if TEST_KCONFIG && RIOT_CI_BUILD
```
Would it be an option to instead have another BOARD that inherits from the original enabling all these extra modules? IMO there should not be dependencies against anything that is not a MODULE, FEATURE (or CPU/CPU_MODEL/etc.)

MrKevinWeiss replied:
Yes that would be another option, the virtual board idea.  I think that would be a different PR then though.  One more board for the CI.  @gschorcht @leandrolanzieri @fjmolinas If we think this is the way forward I would remove the ESP fixes for this PR.

gschorcht replied:
Hm, I already had an approach with two pure virtual board definitions, one for ESP8266 and one for ESP32, which use these optional modules. But this approach was discarded because we tried not to further increase the number of boards and thus the compilation time in Murdock.

gschorcht replied:
Regarding the virtual boards, we had a discussion in PR #12971.

fjmolinas replied:
> Regarding the virtual boards, we had a discussion in PR #12971.

I can't find in the PR the arguments against virtual BOARDs, is there a specific thread? IMO custom dependencies is not great, its easy for those to spiral out of control and for the ordeal to become messy as it was in the past, I would probably have blocked #17314 because of this.

Personally if build is an issue what I would have had is to add to the `externals-boards` test two clones of `esp32` and `esp86` with all those modules enabled. Then you could simply add those external-boards in the required test application. I opened https://github.com/RIOT-OS/RIOT/pull/17525 as a demonstrator for that.

MrKevinWeiss commented:
Dropped changes to the esp

leandrolanzieri commented on Kconfig:
```diff
@@ -43,6 +43,12 @@ comment "Some configuration options may not be here. Use CFLAGS instead."
 comment "!! ERROR: There are conflicting modules active !!"
     depends on ERROR_MODULES_CONFLICT != ""
 
+config RIOT_CI_BUILD
+    bool "CI build"
+    help
+        Indicates that the CI is building. This may add additional modules
+        for the sake of testing.
```
What about taking it from the environment?
```
config RIOT_CI_BUILD
    bool
    default y if '$(RIOT_CI_BUILD)' = '1'
    help
        Indicates that the CI is building. This may add additional modules
        for the sake of testing.
```

leandrolanzieri commented on makefiles/kconfig.mk:
```diff
@@ -121,6 +121,11 @@ ifeq (1,$(DEVELHELP))
 endif
 
+# Expose RIOT_CI_BUILD to kconfig
+ifeq (1,$(RIOT_CI_BUILD))
```
With the above this would not be needed.

leandrolanzieri commented on boards/esp32-wrover-kit/Kconfig:
```diff
@@ -22,5 +22,12 @@ config BOARD_ESP32_WROVER_KIT
     select HAS_SDCARD_SPI
 
     select HAVE_ILI9341
+    select HAVE_RIOT_CI_BUILD_ESP_HW_COUNTER
+    select HAVE_RIOT_CI_BUILD_ESP_I2C_HW
+    select MODULE_ESP_IDF_HEAP if TEST_KCONFIG && RIOT_CI_BUILD
+    select MODULE_ESP_LOG_STARTUP if TEST_KCONFIG && RIOT_CI_BUILD
+    select MODULE_ESP_LOG_TAGGED if TEST_KCONFIG && RIOT_CI_BUILD
+    select MODULE_ESP_QEMU if TEST_KCONFIG && RIOT_CI_BUILD
+    select MODULE_ESP_SPI_RAM if TEST_KCONFIG && RIOT_CI_BUILD
```
I thought we were getting rid of these?

MrKevinWeiss replied:
I thought we were too.. I guess my dropped commit didn't contain all the changes... sorry.

MrKevinWeiss commented:
OK looks better, should not contain any conflicting esp changes.

leandrolanzieri commented on drivers/mpu9x50/Kconfig:
```diff
@@ -5,30 +5,44 @@
-choice
-    bool "MPU-9X50 accelerometer/magnetometer/gyroscope"
-    optional
+menuconfig MODULE_MPU9X50
```
I think you are missing the `CONFIG_MODULE_MPU9X50=y` in the `app.config` of the test application.

MrKevinWeiss replied:
Hmm I wonder if we want it like that or for the entry point to be `MODULE_MPU9150` only?

MrKevinWeiss commented:
I think everything should be in order now.  Any remaining issues?

MrKevinWeiss commented:
ugg I guess this should be paused until the design patterns PR is finished.

MrKevinWeiss commented:
The issue is how to have default selectable modules that don't contain variants (ie no choices).  This currently will hide the prompt during `menuconfig` if the default is selected.  The 2 proposed solutions are selecting from the HAVE_ which has the problem that all dependencies need to be repeated there, imply would not actually updated the values if the default is selected. Or adding a comment, which takes a few extra lines but seems to be the cleanest solution.

MrKevinWeiss commented:
OK, simplified the drivers as per our best understanding of the design patterns.

If everything looks good may I squash?

leandrolanzieri commented on drivers/adxl345/Kconfig:
```diff
@@ -5,8 +5,14 @@
 #
 
-config MODULE_ADXL345
+menuconfig MODULE_ADXL345
```
It has no configurations inside
```suggestion
config MODULE_ADXL345
```

MrKevinWeiss replied:
I set everything to menuconfig as adding things later would be a bit cleaner and I don't know if any functional difference.

If you feel strongly about it I will revert everything to `config` only.

leandrolanzieri commented on drivers/mpu9x50/Kconfig:
```diff
@@ -5,30 +5,44 @@
-        This driver supports both MPU9150 and MPU9250. Choose one model.
+    select MODULE_PERIPH_I2C
+    select MODULE_ZTIMER
+    select MODULE_ZTIMER_MSEC
```
I'd keep the `help` string.

MrKevinWeiss replied:
I added it back in but I somehow think it is equivalent to:
![image](https://user-images.githubusercontent.com/19396439/151961639-d4fda5ae-d74b-4ae9-91ef-a6ace6c4eda0.png)

leandrolanzieri replied:
Think about generating documentation from Kconfig symbols in the future 😉

leandrolanzieri reviewed (commented):
Looks mostly good, please address the comments and squash directly @MrKevinWeiss!

MrKevinWeiss commented:
Just needs an ack :)

fjmolinas reviewed (approved):
LGTM, trusting CI

leandrolanzieri reviewed (approved):
ACK!

MrKevinWeiss commented:
Thanks for the review!
